### PR TITLE
Feature rewards system

### PR DESCRIPTION
**Pull request by krogers78**
This feature will allow users to earn points for chatting on the server, general activity like streaming, gaming, etc, and inviting new users to join the server. Along with that, admins are able to use a command to return a list of users with the most points sorted highest to lowest.

### CODE REVIEW

**Review by MikeSewell — CHANGES_REQUESTED** [review]
There should be some documentation letting the user know about the new rewards system.

**Comment by krogers78** [comment]
Looks good to me! Love the addition of the usage for the new feature.

**Comment by BHess2653** [comment]
Looks good to me

**Comment by drubiodev** [comment]
Looks good

**Review by MikeSewell — APPROVED** [review]

File: bot/client.js [hunk 1]
```diff
@@ -10,6 +11,39 @@ const client = new Discord.Client();
...
+  messagesCountTemp === '000' ? points = currentPoints + amountOfBonusPoints : null;
+  return points;
+};
+
+// Function to add points for chatting
+const awardPointsforChatting = async (message) => {
+  const { content, channel, author } = message;
+  if (channel.type !== 'dm' && content.length >= 5) {
+    const messagesPoints = 0.2;
+    const memberData = await Member.findAll({
```
Comment (reactivepixel) [inline bot/client.js]: Over usage of await

File: bot/client.js [hunk 2]
```diff
@@ -54,7 +88,18 @@ client.on('message', (message) => {
...
       message.reply(helpString);
     }
+  } else {
+    // Award points if the message isn't a command
+    awardPointsforChatting(message);
   }
 });
 
+// When user plays, streams, or Listens give 5 points
+// oldMember is required due to documentation
```
Comment (reactivepixel) [inline bot/client.js:68]: Add a url to the specific docs that are an issue.

File: bot/client.js [hunk 3]
```diff
@@ -54,7 +88,18 @@ client.on('message', (message) => {
...
+client.on('presenceUpdate', (oldMember) => {
+  const { id } = oldMember.user;
+  const { game } = oldMember.presence;
+  !game ? getUserPointsandUpdate(id, 5) : null;
```
Comment (reactivepixel) [inline bot/client.js]: Magic number has no context

File: bot/controllers/invite.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        'Receive one bonus point for every Full Sail student you invite that can be verified.',
+        this.inviteAction.bind(controller),
+        'dm',
+      ),
+    ];
+  }
+  // this message will be sent to the user's dm with their username
+  inviteAction() {
+    const { message } = this;
+    const validDomains = ['student.fullsail.edu', 'fullsail.edu', 'fullsail.com'];
```
Comment (reactivepixel) [inline bot/controllers/invite.js]: Not dry add to the util.

File: bot/controllers/invite.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    const email = message.parsed[1].toLowerCase();
```
Comment (reactivepixel) [inline bot/controllers/invite.js]: Validate the email address is valid

File: bot/controllers/invite.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    const emailDomain = email.split('@').pop();
+    // create the invite for the channel
+    message.channel.createInvite().then((invite, err) => {
+      if (err)util.log('create invite err : ', err, 3);
```
Comment (reactivepixel) [inline bot/controllers/invite.js]: Should abort processing of this function if err

File: bot/controllers/invite.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+          if (err) {
+            message.reply(errorMsg);
+            util.log('Email not sent', err, 3);
+            message.author.send('Invite not sent ');
+          } else {
+            util.log('Email details', info, 3);
+            // for every invite sent the user will receive 1 point
+            getUserPointsandUpdate(message.author.id, 1);
+            // notify the user that the message has been sent
+            message.author.send('you have sent the invite');
```
Comment (reactivepixel) [inline bot/controllers/invite.js]: Sentences start with Capitals and end with punctuation.

File: bot/controllers/invite.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+          }
+        });
+      } else {
+        message.author.send('Sorry the invite could not be sent please contact a admin for assistance');
+      }
+    });
+    return 'You have received 1 point';
```
Comment (reactivepixel) [inline bot/controllers/invite.js]: Variable, not a string of 1

File: bot/controllers/topUsers.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+const { Member } = require('../../db/models');
+
+class topPointsController extends BaseController {
+  constructor(message) {
+    super(message);
+    const controller = this;
+    this.commands = [
+      new Command(
+        '!topUsers',
+        '!topUsers <Limit_Wanted>',
```
Comment (reactivepixel) [inline bot/controllers/topUsers.js]: Consider another phrase, number_of_users

File: bot/controllers/topUsers.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    // Check to see if the user added a number
+    if (!isNaN(message.parsed[1])) {
+      // Grab all the entries with the limit and sort descending
+      Member.findAll({
+        where: { verified: 1 },
+        limit: parseInt(message.parsed[1], 10),
+        order: [['points', 'DESC']],
+      }).then((results) => {
+        // Map out the data and set it equal to a variable
+        const fullMessage = results.map(element =>
```
Comment (reactivepixel) [inline bot/controllers/topUsers.js]: element is generic

File: bot/controllers/verify.js [hunk 11]
```diff
@@ -58,6 +58,8 @@ class VerifyController extends BaseController {
               email,
               uuid: uuidv4(),
               verified: 1,
+              messagesCount: 0,
+              points: 0,
```
Comment (reactivepixel) [inline bot/controllers/verify.js]: Check existing logic for what happens in the event a user double verifies.

Ideally you don't want to wipe points out due to user ignorance.

**Review by reactivepixel — CHANGES_REQUESTED** [review]

**Comment by krogers78** [comment]
Hey guys, I'm going to remove any trace of the messagesCount, add the welcome command, and create a `isVerified` util function.

**Comment by LevanceWam** [comment]
Hey guys I am going to move the valid domains, make spelling and punctuation , and validating that the email is valid.

File: bot/client.js (hunk 3 above)
Comment (BHess2653) [reply bot/client.js]: I will update this with a link to the docs about oldMember and newMember and give context to the magic number.

**Comment by BHess2653** [comment]
Hey guys I am going to add context to the magic number and add docs to oldMember at the bottom of client.js

File: bot/controllers/topUsers.js (hunk 9 above)
Comment (krogers78) [reply bot/controllers/topUsers.js]: I will make sure to give this more context!

File: bot/controllers/topUsers.js (hunk 10 above)
Comment (krogers78) [reply bot/controllers/topUsers.js]: Definitely will give this a new name to better suit the data.

File: bot/controllers/verify.js (hunk 11 above)
Comment (krogers78) [reply bot/controllers/verify.js]: After checking the logic in this file, trying to verify for a second time will not reset their points. Logic is in place to ensure that users cannot be created twice.

File: bot/client.js (hunk 2 above)
Comment (krogers78) [reply bot/client.js:68]: This has been added although the comment has not collapsed.

**Pull request closed without merging**